src/network/network_gui.cpp

```diff
@@ -1564,333 +1575,738 @@ NetworkCompanyInfo *GetLobbyCompanyInfo(CompanyID company)
+	SpriteID sprite;   ///< The sprite to use on the button.
+	StringID tooltip;  ///< The tooltip of the button.
+	Colours colour;    ///< The colour of the button.
+	bool disabled;     ///< Is the button disabled?
+	uint height;       ///< Calculated height of the button.
+	uint width;        ///< Calculated width of the button.
+
+	ButtonCommon(SpriteID sprite, StringID tooltip, Colours colour, bool disabled = false) :
+		sprite(sprite),
+		tooltip(tooltip),
+		colour(colour),
+		disabled(disabled)
+	{
+		Dimension d = GetSpriteSize(sprite);
+		this->height = d.height + ScaleGUITrad(WD_FRAMERECT_TOP + WD_FRAMERECT_BOTTOM);
+		this->width = d.width + ScaleGUITrad(WD_FRAMERECT_LEFT + WD_FRAMERECT_RIGHT);
+	}
+	virtual ~ButtonCommon() {}
+
+	virtual void OnClick(struct NetworkClientListWindow *w, Point pt) = 0;
+};
+
+template<typename T>
+class Button : public ButtonCommon {
+private:
+	typedef void (*ButtonCallback)(struct NetworkClientListWindow *w, Point pt, T id);
+	T id;                 ///< ID this button belongs to.
+	ButtonCallback proc;  ///< Callback proc to call when button is pressed.
+
+public:
+	Button(SpriteID sprite, StringID tooltip, Colours colour, T id, ButtonCallback proc, bool disabled = false) :
+		ButtonCommon(sprite, tooltip, colour, disabled),
+		id(id),
+		proc(proc)
+	{
+		assert(proc != nullptr);
+	}
 
-	uint sel_index;
-	ClientID client_id;
-	Point desired_location;
-	std::vector<ClientListAction> actions; ///< Actions to execute
+	void OnClick(struct NetworkClientListWindow *w, Point pt) override
+	{
+		if (this->disabled) return;
+
+		this->proc(w, pt, this->id);
+	}
+};
+
+using CompanyButton = Button<CompanyID>;
+using ClientButton = Button<ClientID>;
+
+/**
+ * Main handle for clientlist
+ */
+struct NetworkClientListWindow : Window {
+private:
+	ClientListWidgets query_widget;
+	CompanyID join_company;
+
+	ClientID dd_client_id;
+	CompanyID dd_company_id;
+
+	Scrollbar *vscroll;
+	uint line_height;
+	uint line_count;
+	int hover_index;
+	int player_self_index;
+	int player_host_index;
+
+	std::map<uint, std::vector<std::unique_ptr<ButtonCommon>>> buttons;
+
+	static const int CLIENT_OFFSET_LEFT = 12; ///< Offset of client entries compared to company entries.
 
 	/**
-	 * Add an action to the list of actions to execute.
-	 * @param name the name of the action
-	 * @param proc the procedure to execute for the action
+	 * Chat button on a Company is clicked.
 	 */
-	inline void AddAction(StringID name, ClientList_Action_Proc *proc)
+	static void OnClickCompanyChat(NetworkClientListWindow *w, Point pt, CompanyID company_id)
 	{
-		this->actions.push_back({name, proc});
+		ShowNetworkChatQueryWindow(DESTTYPE_TEAM, company_id);
 	}
 
-	NetworkClientListPopupWindow(WindowDesc *desc, int x, int y, ClientID client_id) :
-			Window(desc),
-			sel_index(0), client_id(client_id)
+	/**
+	 * Join button on a Company is clicked.
+	 */
+	static void OnClickCompanyJoin(NetworkClientListWindow *w, Point pt, CompanyID company_id)
```

Comment:
Technically speaking, instead of making these static, you could use [std::bind](https://en.cppreference.com/w/cpp/utility/functional/bind) ;)

Reply:
Nope, forget std::bind, it never existed.